### PR TITLE
ASoC: SOF: hda: fix infinite XRUN after reusuming from S3

### PR DESCRIPTION
**Pull request by RanderWang**
BugLink: https://github.com/thesofproject/sof/issues/1594

In ASoC, hda codecs and SOF driver set the same stream tag
to hardware for a stream and this stream tag should be restored
to hardware after resuming from S3. For this bug, there are
two capture streams active before S3 and one is released when
doing S3. After resuming from S3, hda codec driver restores
stream tag for the active stream, but SOF restores another
stream tag which is allocated again, and this makes controller
don't get any capture data, then infinite XRUN happens in FW.

Now this patch checks whether there is a stored dma data for
stream resuming from S3 and reuses it. And it also removes
dma data when the stream is released.

Tested on whiskylake with SOF driver.

It fixes https://github.com/thesofproject/sof/issues/1594

### CODE REVIEW

**Comment by keyonjie** [comment]
Hi @RanderWang are you meaning that for the active stream, we didn't call hda_link_hw_free() before entering S3?

**Comment by RanderWang** [comment]
> Hi @RanderWang are you meaning that for the active stream, we didn't call hda_link_hw_free() before entering S3?

This should not be done by driver. It is maintained by ASoC.
For active stream,  ASoC doesn't invoke any free function when doing S3.

**Comment by keyonjie** [comment]
> > Hi @RanderWang are you meaning that for the active stream, we didn't call hda_link_hw_free() before entering S3?
> 
> This should not be done by driver. It is maintained by ASoC.
> For active stream, ASoC doesn't invoke any free function when doing S3.

If hw_free() is not called before suspend, we shouldn't call hw_params() at resume, isn't it?

**Comment by kv2019i** [comment]
@keyonjie wrote:
> If hw_free() is not called before suspend, we shouldn't call hw_params() at resume, isn't it?

I'm afraid we have to as the DSP has lost its state due to power cycle, so we need to program the link DMA again. ASoC won't call hw_params again as hw_params have not changed, but it's upto us to reprorammg the hardware in prepare().

@RanderWang what I don't understand from the patch is that why hda_link_stream_assign() gets confused -- the functions seems to be written in a way that it can be called multiple times (like hw_params() can) and should return the correct result. I.e. hda_check_fes() should always find a matching result. When you say other stream is removed, hda_check_fes() should still work for the remainin stream, right?

All in all this is a very good find, thanks Rander!

**Comment by kv2019i** [comment]
@RanderWang I have some comments on the commit message:

> ASoC: SOF: hda: fix infinite XRUN after reusuming from S3

The XRUN loop is a second order outcome. How about:

ASoC: SOF: hda: fix link DMA config 

> BugLink: thesofproject/sof#1594

Don't but the buglink at start of git commit. I don't think we have established practise for this.
Pierre has done this a couple of times and the patches have been accepted, so should be ok:

GitHub issue: https://github.com/thesofproject/sof/issues/1594

And put it at end end, in the same section as the signed-off-by line (see Pierre's patches).

> two capture streams active before S3 and one is released when
> doing S3. After resuming from S3, hda codec driver restores

This sounds a little odd. Is the other stream terminated by the user while S3 resume is still not complete? Does the audio file end during S3 resume? 

> Tested on whiskylake with SOF driver.

Please use the correct spelling: "Tested on a Whiskey Lake platform".
https://ark.intel.com/content/www/us/en/ark/products/codename/135883/whiskey-lake.html

As this is a SOF driver patch, no need to say you tested with SOF driver.

**Comment by ranj063** [comment]
Good finding @RanderWang . Please fix typo in the commit message apart from the things @kv2019i mentioned too reusuming -> resuming

**Comment by RanderWang** [comment]
> @RanderWang what I don't understand from the patch is that why hda_link_stream_assign() gets confused -- the functions seems to be written in a way that it can be called multiple times (like hw_params() can) and should return the correct result. I.e. hda_check_fes() should always find a matching result. When you say other stream is removed, hda_check_fes() should still work for the remainin stream, right?
> 

Thanks!  I will refine my explanation.   Actually, there are two pcm capture streams, which is composed of a FE stream and a BE stream. One of the pcm capture stream is released when doing S3. hda_check_fes is still work for another pcm capture stream.

**Comment by RanderWang** [comment]
> 
> > two capture streams active before S3 and one is released when
> > doing S3. After resuming from S3, hda codec driver restores
> 
> This sounds a little odd. Is the other stream terminated by the user while S3 resume is still not complete? Does the audio file end during S3 resume?
> 

I will refine my comments.  The other unused pcm capture stream is terminated by pulse audio and it occurs before S3. The termination is not caused by S3.

**Comment by RanderWang** [comment]
Updated my PR and fix another stream id issue(please check https://github.com/thesofproject/linux/commit/aab9bf9d5d3f45bea4fe932acf64298b410c0495)

@ranj063  Please check the new change which modifies your algorithm for S3.   Thanks!
I also test the case for you patch : play a music ---  pause playback -- do S3 suspend ---do S3 resume ---  continue playback .

**Comment by kv2019i** [comment]
@RanderWang The patch looks good now, but I'm still having difficulty with the commit message.

- You still misspell Whiskey Lake (Whiskey not Whisky)  :D See the link above for the official page.

- This sentence doesn't calculate:

> After system resume, hda codec driver restores the stream tag for
> the active pcm stream, but SOF restores another stream tag which is
> allocated with a different stream tag, and this makes controller

A stream tag allocated with a different stream tag, this doesn't make sense.
How about this wording:

After system resume, hda codec driver restores the stream tag for
the active pcm stream, but SOF goes to assign a new one, which now 
doesn't match with the stream tag used by codec driver, and this makes 
controller...

... I think this is the main issue. What I still don't fully understand is
that why hda_link_stream_assign() doesn't find the right stream_tag
with current code. We are not freeing the 'link_dev' instance, so
the data with the right tag should still be there. Oh well, you don't
need to explain this in the commit -- this is more to understand
whether the fix is right or not.

File: sound/soc/sof/intel/hda-dai.c [hunk 1]
```diff
@@ -330,7 +330,8 @@ static int hda_link_pcm_trigger(struct snd_pcm_substream *substream,
 		if (ret < 0)
 			return ret;
 		stream_tag = hdac_stream(link_dev)->stream_tag;
-		snd_hdac_ext_link_clear_stream_id(link, stream_tag);
+		if (substream->stream == SNDRV_PCM_STREAM_PLAYBACK)
+			snd_hdac_ext_link_clear_stream_id(link, stream_tag);
```
Comment (kv2019i) [inline sound/soc/sof/intel/hda-dai.c:333]: @RanderWang But why? We need link and host DMA for capture as well, why the different treatment?
UPDATE: answering to myself: I was confused by HDAC interface. snd_hdac_ext_link_clear_stream_id() specifically modifies status for _output_ stream as per HDA spec. Nothing in hdac_ext code hints at this, which is confusing. But oh well, this is outside your patch, sorry for the noise. This second patch looks good.

Comment (kv2019i) [reply sound/soc/sof/intel/hda-dai.c:333]: > ... I think this is the main issue. What I still don't fully understand is
> that why hda_link_stream_assign() doesn't find the right stream_tag
> with current code.

@RanderWang I've been trying to reproduce this by on WHL but to no avail. 
While I still don't fully get how stream_assign() is supposed to work,
it would seem your patch is at least the safer option.

It's confusing that stream_assign() basicly returns the first available hdac-ext stream, and if there
are any non-opened streams in the bus (with matching direction), the caller will get a random stream back. The assumption made is that this only happens in non-host-dma cases, but it's very hard to verify from code review this assumption is ok.

Comment (RanderWang) [reply sound/soc/sof/intel/hda-dai.c:333]: @kv2019i  very good finding ! 
the author of this function is @ranj063 ,  but I can share my understanding.
In ASoC, the calling sequence of playback is :   FE stream open --- BE stream hw_param --- FE stream hw_param.  The hdac_ext stream for FE stream is allocated in hda_dsp_pcm_open firstly, and for BE stream is done in hda_link_hw_params later. So in stream_assign, the assumption is if there are FE & BE for a pcm stream, they will be combined with the same tag when searching list from head to tail.

This is not reasonable, and the allocation is not optimized, .e.g 
(1) pcm1 FE open + BE hw_param// both stream tag 2 is allocated for the FE & BE
(2) pcm2 FE open// stream tag 3 is allocated for the FE
(3) pcm1 FE & BE release //both stream tag 2 is released for the FE & BE
(4) pcm2 BE hw_param // stream tag 2 is allocated for this BE for the FE with stream tag 2 is not used now. But the optimized way is allocating stream tag 3 to the BE.

We should check all the streams to find whether there is a stream used by the FE of this BE, if there is none, the BE should be a hostless stream.

My patch is not for this case. I think we should not release stream for BE when doing S3.  the hda stream for FE in SOF driver is still active and the stream tag is still used by hda codec driver, so we should let ASoC to release all the streams in normal way

**Comment by RanderWang** [comment]
> @RanderWang The patch looks good now, but I'm still having difficulty with the commit message.
> 
> * You still misspell Whiskey Lake (Whiskey not Whisky)  :D See the link above for the official page.
> * This sentence doesn't calculate:
> 
> > After system resume, hda codec driver restores the stream tag for
> > the active pcm stream, but SOF restores another stream tag which is
> > allocated with a different stream tag, and this makes controller
> 
> A stream tag allocated with a different stream tag, this doesn't make sense.
> How about this wording:
> 
> After system resume, hda codec driver restores the stream tag for
> the active pcm stream, but SOF goes to assign a new one, which now
> doesn't match with the stream tag used by codec driver, and this makes
> controller...
> 
> ... I think this is the main issue. What I still don't fully understand is
> that why hda_link_stream_assign() doesn't find the right stream_tag
> with current code. We are not freeing the 'link_dev' instance, so
> the data with the right tag should still be there. Oh well, you don't
> need to explain this in the commit -- this is more to understand
> whether the fix is right or not.

@kv2019i  Thanks for your help on commit message. I will study and refine it.

File: sound/soc/sof/intel/hda-dai.c (hunk 1 above)
Comment (RanderWang) [reply sound/soc/sof/intel/hda-dai.c:333]: I don't know whether you know the background of hda dma.  There are two hda dma engines : host dma and link dma. For SOF,  dsp is enabled and hda dma is set to decoupled mode. host dma channel number is not necessary equal to link dma for a pcm stream. Also there is  a case of N host dma linked to M link dma .  But for a pcm with only one FE and BE, it is best to use the same channel number(stream tag) for a hardware requirement( it is a long story...).

File: sound/soc/sof/intel/hda-dsp.c [hunk 2]
```diff
@@ -455,6 +455,11 @@ int hda_dsp_set_hw_params_upon_resume(struct snd_sof_dev *sdev)
 			if (!link)
 				return -EINVAL;
 			stream_tag = hdac_stream(stream)->stream_tag;
+
+			if (hdac_stream(stream)->direction ==
+				SNDRV_PCM_STREAM_CAPTURE)
+				continue;
+
```
Comment (libinyang) [inline sound/soc/sof/intel/hda-dsp.c:446]: Why not use the same format:
```
	if (substream->stream == SNDRV_PCM_STREAM_PLAYBACK)
		snd_hdac_ext_link_clear_stream_id(link, stream_tag);
```

File: sound/soc/sof/intel/hda-dai.c [hunk 3]
```diff
@@ -372,7 +373,8 @@ static int hda_link_hw_free(struct snd_pcm_substream *substream,
 		return -EINVAL;
 
 	stream_tag = hdac_stream(link_dev)->stream_tag;
-	snd_hdac_ext_link_clear_stream_id(link, stream_tag);
+	if (substream->stream == SNDRV_PCM_STREAM_PLAYBACK)
+		snd_hdac_ext_link_clear_stream_id(link, stream_tag);
```
Comment (libinyang) [inline sound/soc/sof/intel/hda-dai.c:378]: maybe it is better
```
if (substream->stream == SNDRV_PCM_STREAM_PLAYBACK) {
    stream_tag = hdac_stream(link_dev)->stream_tag;
    snd_hdac_ext_link_clear_stream_id(link, stream_tag);
}
```
like other codes in the driver.

File: sound/soc/sof/intel/hda-dsp.c (hunk 2 above)
Comment (RanderWang) [reply sound/soc/sof/intel/hda-dsp.c:446]: Because of limitation of 80 char. With the line of if(xxxx),  
the size of line "snd_hdac_ext_link_clear_stream_id(link, stream_tag);" 
 is also more than 80 chars. to make it like:
snd_hdac_ext_link_clear_stream_id(link,
                                                           stream_tag);
It is still also more than 80 chars. I did this is to make it beautiful

File: sound/soc/sof/intel/hda-dai.c (hunk 3 above)
Comment (RanderWang) [reply sound/soc/sof/intel/hda-dai.c:378]: yes, I will refine it.

**Comment by RanderWang** [comment]
update my PR to refine code style suggested by Libin

**Comment by RanderWang** [comment]
Rebased PR, thanks!

**Comment by RanderWang** [comment]
The failed CI test is not caused by this PR.

File: sound/soc/sof/intel/hda-dai.c (hunk 1 above)
Comment (kv2019i) [reply sound/soc/sof/intel/hda-dai.c:333]: Thanks  @RanderWang for the explanation! The patch and commit message now look ok.
What confused me was the documentation (in hdac library) for snd_hdac_ext_link_clear_stream_id(). It doesn't describe that in HDA spec, this property is only valid for outputs.  But yeah. outside this patch.

**Comment by RanderWang** [comment]
Update my commit message, thanks!

**Comment by ranj063** [comment]
@kv2019i are you good with this PR?
